### PR TITLE
feat: add batch process metrics

### PR DESCRIPTION
**Pull request by gy09535**
### What this PR does / why we need it:

fix: https://github.com/apache/apisix/issues/2693

add metric for batch process, such as:  sls logger, tcp logger ,udp logger , http logger, sys logger. 
with this feature, users can use the metrics find how many entries remaining in batch process.
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/prometheus/exporter.lua [hunk 1]
```diff
@@ -45,10 +45,10 @@ local DEFAULT_BUCKETS = { 1, 2, 5, 7, 10, 15, 20, 25, 30, 40, 50, 60, 70,
 local metrics = {}
 
 
-    local inner_tab_arr = {}
-local function gen_arr(...)
-    clear_tab(inner_tab_arr)
 
+local function gen_arr(...)
```
Comment (spacewander) [inline apisix/plugins/prometheus/exporter.lua:49]: We don't need this if we can't share the table.

File: apisix/utils/batch-processor.lua [hunk 2]
```diff
@@ -146,8 +150,19 @@ function batch_processor:push(entry)
...
+        batch_metrics = prometheus.get_prometheus():gauge("batch_process_entries",
+                                                          "batch process remaining entries",
+                                                          {"name", "route_id", "server_addr"})
+    end
+
     local entries = self.entry_buffer.entries
     table.insert(entries, entry)
+    -- add batch metric for every route
+    if batch_metrics  then
+        batch_metrics:set(#entries, prometheus.gen_arr(self.name, self.route_id, self.server_addr))
```
Comment (spacewander) [inline apisix/utils/batch-processor.lua:164]: We can store the table: `self.xxx = {self.name, self.route_id, self.server_addr}`.

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by Firstsawyou** [comment]
The test case reported an error and we need to resolve it. ^ _ ^

File: apisix/plugins/prometheus/exporter.lua (hunk 1 above)
Comment (gy09535) [reply apisix/plugins/prometheus/exporter.lua:49]: got it.

File: apisix/utils/batch-processor.lua (hunk 2 above)
Comment (gy09535) [reply apisix/utils/batch-processor.lua]: fixed.

File: apisix/utils/batch-processor.lua [hunk 3]
```diff
@@ -146,8 +150,20 @@ function batch_processor:push(entry)
...
+        batch_metrics = prometheus.get_prometheus():gauge("batch_process_entries",
+                                                          "batch process remaining entries",
+                                                          {"name", "route_id", "server_addr"})
+    end
+
     local entries = self.entry_buffer.entries
     table.insert(entries, entry)
+    -- add batch metric for every route
+    if batch_metrics  then
+        local label = {self.name, self.route_id, self.server_addr}
```
Comment (spacewander) [inline apisix/utils/batch-processor.lua:164]: Look like we can store the `label` in `self`?

File: t/plugin/prometheus.t [hunk 4]
```diff
@@ -1088,3 +1088,207 @@ GET /apisix/prometheus/metrics
...
+--- error_code: 404
+
+
+
+=== TEST 64: check sys logger metrics
+--- request
+GET /apisix/prometheus/metrics
+--- error_code: 200
+--- response_body_like eval
+qr/apisix_batch_process_entries{name="sys-logger",route_id="9"/
```
Comment (spacewander) [inline t/plugin/prometheus.t:1176]: Escape `{`?

File: apisix/utils/batch-processor.lua (hunk 3 above)
Comment (gy09535) [reply apisix/utils/batch-processor.lua]: fixed.

File: apisix/utils/batch-processor.lua [hunk 5]
```diff
@@ -146,8 +150,20 @@ function batch_processor:push(entry)
         return
     end
 
+    if not batch_metrics and prometheus.get_prometheus() and self.name
+       and self.route_id and self.server_addr then
+        batch_metrics = prometheus.get_prometheus():gauge("batch_process_entries",
+                                                          "batch process remaining entries",
```
Comment (spacewander) [inline apisix/utils/batch-processor.lua:156]: We need to update the https://github.com/apache/apisix/blob/master/doc/plugins/prometheus.md#available-metrics

File: t/plugin/prometheus.t (hunk 4 above)
Comment (gy09535) [reply t/plugin/prometheus.t]: plugin set error 😄

File: apisix/utils/batch-processor.lua (hunk 5 above)
Comment (gy09535) [reply apisix/utils/batch-processor.lua:156]: added.

File: doc/plugins/prometheus.md [hunk 6]
```diff
@@ -125,6 +125,7 @@ Or you can goto [Grafana official](https://grafana.com/grafana/dashboards/11719)
 * `Bandwidth`: Total Bandwidth (egress/ingress) flowing through apisix. This metric is available per service and as a sum across all services.
 * `etcd reachability`: A gauge type with a value of 0 or 1, representing if etcd can be reached by a apisix or not.
 * `Connections`: Various Nginx connection metrics like active, reading, writing, and number of accepted connections.
+* `Batch process entries`: A gauge type, when we use plugins such as: sys logger, http logger, sls logger, tcp logger, udp logger and zipkin, the surplus entries which not sended will be statistics in the metrics.
```
Comment (spacewander) [inline doc/plugins/prometheus.md:128]: Should mention the batch processor.
Some suggestions:
`surplus entries` => `entries`
`not sended` => `hasn't been sent`
`statistics` => `counted`

Comment (gy09535) [reply doc/plugins/prometheus.md]: changed.

**Review by spacewander — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**